### PR TITLE
Ignore Wrong TileHeader

### PR DESCRIPTION
**Pull request by mackaiver**
Some fits files have a kaputt tile header in the heap containing the DrsCellOffsets. This ignores this problem.

### CODE REVIEW

**Comment by dneise** [comment]
Its just that? its kaput?

**Comment by mackaiver** [comment]
I'm not sure. The old fitsreader implementation ignored the faulty header by default. If I understand the file format 'specification' correctly the "TILE" keyword should be there. And it is for some files. For some it isn't.

File: pom.xml [hunk 1]
```diff
@@ -4,7 +4,7 @@
   <groupId>de.sfb876</groupId>
   <artifactId>fact-tools</artifactId>
   <name>fact-tools</name>
-  <version>0.18.0</version>
+  <version>0.18.1</version>
```
Comment (MaxNoe) [inline pom.xml]: We did not yet release 0.18.0. So this should be also 0.18.0.

File: src/main/java/fact/io/hdureader/FITSStream.java [hunk 2]
```diff
@@ -43,7 +43,9 @@ public void init() throws Exception {
         //get calibration constants they are stored in the first (and only) row of this hdu.
         HDU offsetHDU = fits.getHDU("ZDrsCellOffsets");
         if (offsetHDU != null){
-            OptionalTypesMap<String, Serializable> row = ZFITSHeapReader.forTable(offsetHDU.getBinTable()).getNextRow();
+            ZFITSHeapReader reader = ZFITSHeapReader.forTable(offsetHDU.getBinTable());
+            boolean ignoreWrongTileHeader;
```
Comment (MaxNoe) [inline src/main/java/fact/io/hdureader/FITSStream.java]: Why is it necessary to declare this variable here?

Comment (mackaiver) [inline src/main/java/fact/io/hdureader/FITSStream.java]: It's not. But I find it a whole lot more readable.

Comment (dneise) [inline src/main/java/fact/io/hdureader/FITSStream.java]: Ah this mimics keyword arguments like e.g. in Python, right?

**Comment by dneise** [comment]
> And it is for some files. For some it isn't.

Could you give an example file?

File: src/main/java/fact/io/hdureader/FITSStream.java (hunk 2 above)
Comment (mackaiver) [inline src/main/java/fact/io/hdureader/FITSStream.java]: Yes. If I was to read this code I would find a random `true` argument to that  function very surprising. I think the variable name helps.

**Comment by mackaiver** [comment]
Yes. 20131105_184 for example. And also others from that month at least.

**Comment by mackaiver** [comment]
# Test File
https://github.com/fact-project/fact-tools/blob/master/src/test/resources/testDataFile.fits.fz
(20160817_16)
Header entries for HDU with name ZDrsCellOffset


```
result = {HashMap@1412}  size = 25
 0 = {HashMap$Node@1415} "ZFORM1" -> "1474560I"
 1 = {HashMap$Node@1416} "THEAP" -> "16"
 2 = {HashMap$Node@1417} "CHECKSUM" -> "KEdPN9cNKCcNK9cN"
 3 = {HashMap$Node@1418} "TTYPE1" -> "OffsetCalibration"
 4 = {HashMap$Node@1419} "NAXIS" -> "2"
 5 = {HashMap$Node@1420} "ZTABLE" -> "T"
 6 = {HashMap$Node@1421} "PCOUNT" -> "1383872"
 7 = {HashMap$Node@1422} "RAWSUM" -> "3842390352"
 8 = {HashMap$Node@1423} "TFORM1" -> "1QB"
 9 = {HashMap$Node@1424} "DATASUM" -> "3302774089"
 10 = {HashMap$Node@1425} "ZSHRINK" -> "1"
 11 = {HashMap$Node@1426} "ZCTYP1" -> "FACT"
 12 = {HashMap$Node@1427} "ZRATIO" -> "2.13106417655945"
 13 = {HashMap$Node@1428} "ZHEAPPTR" -> "16"
 14 = {HashMap$Node@1429} "TFIELDS" -> "1"
 15 = {HashMap$Node@1430} "ZPCOUNT" -> "0"
 16 = {HashMap$Node@1431} "EXTNAME" -> "ZDrsCellOffsets"
 17 = {HashMap$Node@1432} "NAXIS2" -> "1"
 18 = {HashMap$Node@1433} "GCOUNT" -> "1"
 19 = {HashMap$Node@1434} "NAXIS1" -> "16"
 20 = {HashMap$Node@1435} "BITPIX" -> "8"
 21 = {HashMap$Node@1436} "XTENSION" -> "BINTABLE"
 22 = {HashMap$Node@1437} "ZNAXIS2" -> "1"
 23 = {HashMap$Node@1438} "ZTILELEN" -> "1"
 24 = {HashMap$Node@1439} "ZNAXIS1" -> "2949120"
 
 ```

Bytes after HDU header with name ZDrsCellOffsets for test file

```
result = "�TILE��R��e"
 value = {char[64]@1544}
  0 = '\u0000' 0
  1 = '\u0000' 0
  2 = '\u0000' 0
  3 = '\u0000' 0
  4 = '\u0000' 0
  5 = '\u0015' 21
  6 = '\u001D' 29
  7 = '�' 65533
  8 = '\u0000' 0
  9 = '\u0000' 0
  10 = '\u0000' 0
  11 = '\u0000' 0
  12 = '\u0000' 0
  13 = '\u0000' 0
  14 = '\u0000' 0
  15 = '\u0010' 16
  16 = 'T' 84        <---------- this is correct. Tile starts 16 bytes after the header.
  17 = 'I' 73
  18 = 'L' 76
  19 = 'E' 69
  20 = '\u0001' 1
  21 = '\u0000' 0
  22 = '\u0000' 0
  23 = '\u0000' 0
  24 = '�' 65533
  25 = '\u001D' 29
  26 = '\u0015' 21
  27 = '\u0000' 0
  28 = '\u0000' 0
  29 = '\u0000' 0
  30 = '\u0000' 0
  31 = '\u0000' 0
  32 = '�' 65533
  33 = '\u001D' 29
  34 = '\u0015' 21
  35 = '\u0000' 0
  36 = '\u0000' 0
  37 = '\u0000' 0
  38 = '\u0000' 0
  39 = '\u0000' 0
  40 = 'R' 82
  41 = '\u0001' 1
  42 = '\u0002' 2
  43 = '\u0000' 0
  44 = '�' 65533
  45 = '\u001D' 29
  46 = '\u0015' 21
  47 = '\u0000' 0
  48 = '\u0000' 0
  49 = '�' 65533
  50 = '\u0016' 22
  51 = '\u0000' 0
  52 = '\u0000' 0
  53 = '\u0000' 0
  54 = '\u0000' 0
  55 = '\u0000' 0
  56 = 'e' 101
  57 = '\u0001' 1
  58 = '\u0000' 0
  59 = '\u0000' 0
  60 = '\u0000' 0
  61 = '\u0000' 0
  62 = '\u0000' 0
  63 = '\u0000' 0

```

# Old File

20131105_184.fits.fz

Header entries for HDU with name ZDrsCellOffset
```
result = {HashMap@1302}  size = 25
 0 = {HashMap$Node@1305} "ZFORM1" -> "1474560I"
 1 = {HashMap$Node@1306} "THEAP" -> "16"
 2 = {HashMap$Node@1307} "CHECKSUM" -> "gA8Vh45TgA5Tg35T"
 3 = {HashMap$Node@1308} "TTYPE1" -> "OffsetCalibration"
 4 = {HashMap$Node@1309} "NAXIS" -> "2"
 5 = {HashMap$Node@1310} "ZTABLE" -> "T"
 6 = {HashMap$Node@1311} "PCOUNT" -> "1381250"
 7 = {HashMap$Node@1312} "RAWSUM" -> "4040282208"
 8 = {HashMap$Node@1313} "TFORM1" -> "1QB"
 9 = {HashMap$Node@1314} "DATASUM" -> "3918279670"
 10 = {HashMap$Node@1315} "ZSHRINK" -> "1"
 11 = {HashMap$Node@1316} "ZCTYP1" -> "FACT"
 12 = {HashMap$Node@1317} "ZRATIO" -> "2.13513422012329"
 13 = {HashMap$Node@1318} "ZHEAPPTR" -> "16"
 14 = {HashMap$Node@1319} "TFIELDS" -> "1"
 15 = {HashMap$Node@1320} "ZPCOUNT" -> "0"
 16 = {HashMap$Node@1321} "EXTNAME" -> "ZDrsCellOffsets"
 17 = {HashMap$Node@1322} "NAXIS2" -> "1"
 18 = {HashMap$Node@1323} "GCOUNT" -> "1"
 19 = {HashMap$Node@1324} "NAXIS1" -> "16"
 20 = {HashMap$Node@1325} "BITPIX" -> "8"
 21 = {HashMap$Node@1326} "XTENSION" -> "BINTABLE"
 22 = {HashMap$Node@1327} "ZNAXIS2" -> "1"
 23 = {HashMap$Node@1328} "ZTILELEN" -> "1"
 24 = {HashMap$Node@1329} "ZNAXIS1" -> "2949120"
 ```
 

Bytes after HDU header with name ZDrsCellOffsets for file 20131105_184.fits.fz

```
result = "brbRR�g"
 value = {char[64]@1398}
  0 = '\u0000' 0
  1 = '\u0000' 0
  2 = '\u0000' 0
  3 = '\u0000' 0
  4 = '\u0000' 0
  5 = '\u0015' 21
  6 = '\u0013' 19
  7 = 'b' 98
  8 = '\u0000' 0
  9 = '\u0000' 0
  10 = '\u0000' 0
  11 = '\u0000' 0
  12 = '\u0000' 0
  13 = '\u0000' 0
  14 = '\u0000' 0
  15 = '\u0010' 16
  16 = '\u0000' 0
  17 = '\u0000' 0
  18 = '\u0000' 0
  19 = '\u0000' 0
  20 = '\u0001' 1
  21 = '\u0000' 0
  22 = '\u0000' 0
  23 = '\u0000' 0
  24 = 'r' 114
  25 = '\u0013' 19
  26 = '\u0015' 21
  27 = '\u0000' 0
  28 = '\u0000' 0
  29 = '\u0000' 0
  30 = '\u0000' 0
  31 = '\u0000' 0
  32 = 'b' 98
  33 = '\u0013' 19
  34 = '\u0015' 21
  35 = '\u0000' 0
  36 = '\u0000' 0
  37 = '\u0000' 0
  38 = '\u0000' 0
  39 = '\u0000' 0
  40 = 'R' 82
  41 = '\u0001' 1
  42 = '\u0002' 2
  43 = '\u0000' 0
  44 = 'R' 82
  45 = '\u0013' 19
  46 = '\u0015' 21
  47 = '\u0000' 0
  48 = '\u0000' 0
  49 = '�' 65533
  50 = '\u0016' 22
  51 = '\u0000' 0
  52 = '\u0000' 0
  53 = '\u0000' 0
  54 = '\u0000' 0
  55 = '\u0000' 0
  56 = 'g' 103
  57 = '\u0001' 1
  58 = '\u0000' 0
  59 = '\u0000' 0
  60 = '\u0000' 0
  61 = '\u0000' 0
  62 = '\u0000' 0
  63 = '\u0000' 0
 hash = 0
 
 ```

**Comment by dneise** [comment]
at what event is the TILE missing or rather at what address in that "heap"?

**Comment by dneise** [comment]
This close button is simply at the wrong position...

**Comment by mackaiver** [comment]
Its not an 'event' its the HDU with the ZDrsCellOffsets. So if I'm not mistaken the word `TILE` should be the first heap after the two header. PrimaryHeader + SecondHeader +   16 = 2880 + 2880 + 16.

**Comment by mackaiver** [comment]
Thats byte 5776

File: src/main/java/fact/io/hdureader/FITSStream.java (hunk 2 above)
Comment (MaxNoe) [inline src/main/java/fact/io/hdureader/FITSStream.java]: I was just confused, what is happening in java, if you assign something inside a function call? 

I expected this:
```java
boolean ignoreWrongTileHeader = true;
OptionalTypesMap<String, Serializable> row = reader.getNextRow(ignoreWrongTileHeader);
```

This assignment inside the function call confused me.

**Pull request merged**